### DISCUSSION TITLE
[ideas] save and read targets as on disk formats

### DISCUSSION THREAD

**Discussion by stemangiola** [ropensci/targets#1278] [Ideas]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I see a `format = "parquet"` but when I `tar_read`all data is loaded into memory, and crashed the session (> 300Tb). 

I was wondering if the was a way to save and read

- data.frame in parquet
- format= HDF5 gene expression as on-disk H5

**Comment by multimeric** [comment]
As described [here](https://docs.ropensci.org/targets/reference/tar_target.html#storage-formats), `format = "parquet"` tells Targets to use [`arrow::write_parquet()`](https://arrow.apache.org/docs/r/reference/write_parquet.html) and [`arrow::read_parquet()`](https://arrow.apache.org/docs/r/reference/read_parquet.html). Unfortunately those functions don't actually exploit the parquet format's ability to be queried directly on disk. 

My workaround for this type of thing has been to treat the parquet as a file (ie `format = "file"`, and I return the path to that file from my target), but this is admittedly not taking advantage of Targets' abstractions. I think the better idea would be to define a custom format using [`tar_format`](https://docs.ropensci.org/targets/reference/tar_format.html) that saves to parquet using `arrow` but that loads it using something like `duckdb`/`dbplyr`/`duckplyr` to take advantage of the disk backing.

**Reply by stemangiola** [reply]
that would be a game changer! As well as for Anndata and/or SingleCellExperiment H5.

I give you an example where this abstraction would be awesome

I am using `tar_group`, which is simply adding a group column to a target data frame, however that target data frame includes tibbles (in a data column) that together are very heavy, more than 100 GB. this trivial operation requires now a big virtual machine and a lot of time rather than requiring a very little machine and little time. For this operation, the data does not need to be loaded at all.

Would you have any interest in implementing it?

**Reply by multimeric** [reply]
I'm not certain that disk backing would help here, because as soon as you modify the data frame, the whole thing needs to be loaded from disk in order written to disk.

**Reply by stemangiola** [reply]
I mean the target `foo` is

```
dataframe
study_id | data<parquet, on disk> 
1              | ...
1              | ...
2             | ...
.........
...               | ...
```

And I want to use `tar_group(foo_grouped, foo, study_id)`

and `targets` would transform it into

```
study_id | data<parquet, on disk>  | tar_group
A              | ...                                     | 1
A              | ...                                     | 1
B             | ...                                     | 2
.........
...               | ...                                    |...
```

But `targets` never needs to read the column `data`.

**Reply by stemangiola** [reply]
Same thing If I want to `tar_read` a huge list of tibbles, and just select few items, instead of needing 300Gb of memory just to read the list, I would use little in case I just need few elements.

**Reply by multimeric** [reply]
I think that your targets should return a data frame with `data` as a character vector of filepaths to the parquet files. Then when you need the actual data you can load the parquet as needed (with `dbplyr` etc). I'm not sure we can expect Targets to do what you want here automatically. Nesting a disk-backed data frame inside another data frame is a bit complex.

**Reply by stemangiola** [reply]
I see. I will think about it, having files outside the targets framework can have its drawbacks. @wlandau any opinion? How hard would be to implement native on-disk file-system?

**Reply by wlandau** [reply]
Natively, it doesn't seem feasible because of the number of things in `targets` which assume a whole target is loaded at a time. But there could be a `tar_format()`-powered workaround in which each saved file is just a reference to a piece of a large data file, as described in https://github.com/ropensci/targets/discussions/1232#discussioncomment-9024232 for a different situation.

**Reply by wlandau** [reply]
You might also try https://github.com/ropensci/targets/discussions/1281#discussioncomment-9437279
